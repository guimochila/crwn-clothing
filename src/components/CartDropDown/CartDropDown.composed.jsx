import React from 'react';
import { withRouter } from 'react-router-dom';
import { Query, Mutation } from 'react-apollo';
import { adopt } from 'react-adopt';

import CartDropDown from './CartDropDown';
import { TOGGLE_CART_HIDDEN, GET_CART_ITEMS } from '../../graphql/queries';

const Composed = adopt({
  toggleCartHidden: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_HIDDEN}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={GET_CART_ITEMS}>{render}</Query>,
});

const CartDropDownComposed = props => (
  <Composed>
    {({ toggleCartHidden, localState }) => {
      const { cartItems } = localState.data;

      return (
        <CartDropDown
          toggleCartHidden={toggleCartHidden}
          cartItems={cartItems}
          {...props}
        />
      );
    }}
  </Composed>
);

export default withRouter(CartDropDownComposed);
