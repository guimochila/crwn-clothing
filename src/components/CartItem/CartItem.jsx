import React from 'react';
import {
  CartItemContainer,
  CartItemImage,
  ItemDetailsContainer,
  ItemName,
} from './CartItem.styles';

function CartItem({ item: { imageUrl, price, name, quantity } }) {
  return (
    <CartItemContainer>
      <CartItemImage src={imageUrl} alt={name} />
      <ItemDetailsContainer>
        <ItemName>{name}</ItemName>
        <span>
          {quantity} x ${price}
        </span>
      </ItemDetailsContainer>
    </CartItemContainer>
  );
}

export default React.memo(CartItem);
