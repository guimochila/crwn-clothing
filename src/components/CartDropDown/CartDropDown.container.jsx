import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';

import CartDropDown from './CartDropDown';
import { selectCartItems } from '../../store/cart/cart.selectors';

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

const CartDropDownContainer = compose(
  withRouter,
  connect(mapStateToProps),
)(CartDropDown);

export default CartDropDownContainer;
