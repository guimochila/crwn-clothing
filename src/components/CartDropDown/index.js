import CartDropDown from './CartDropDown';
import CartDropDownContainer from './CartDropDown.container';

export { CartDropDown, CartDropDownContainer };
