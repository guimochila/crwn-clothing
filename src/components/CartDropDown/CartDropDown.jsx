import React from 'react';

import CartItem from '../CartItem';
import {
  CartDropDownContainer,
  CartItemsContainer,
  CartDropDownButton,
  EmptyMessageContainer,
} from './CartDropDown.styles';

function CartDropDown({ cartItems, history, toggleCartHidden }) {
  return (
    <CartDropDownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map(cartItem => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
        )}
      </CartItemsContainer>
      <CartDropDownButton
        onClick={() => {
          history.push('/checkout');
          toggleCartHidden();
        }}
      >
        GO TO CHECKOUT
      </CartDropDownButton>
    </CartDropDownContainer>
  );
}

export default CartDropDown;
