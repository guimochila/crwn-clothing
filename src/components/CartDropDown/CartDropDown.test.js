import React from 'react';
import { shallow } from 'enzyme';

import CartDropDown from './CartDropDown';
import CartItem from '../CartItem';

import { toggleCartHidden } from '../../store/cart/cart.actions';

describe('CardDropDown component', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      dispatch: mockDispatch,
      history: mockHistory,
    };

    wrapper = shallow(<CartDropDown {...mockProps} />);
  });

  test('should render CardDropDown', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call history.push when button is clicked', () => {
    wrapper.find('CartDropDownButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });

  test('should render the number of CartItems components equal the cartItem props', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  test('should render EmptyMessage when cartItems is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch,
    };
    const newWrapper = shallow(<CartDropDown {...mockProps} />);
    expect(newWrapper.exists('EmptyMessageContainer')).toBe(true);
  });
});
