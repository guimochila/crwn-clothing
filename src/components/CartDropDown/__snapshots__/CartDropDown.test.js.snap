// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDropDown component should render CardDropDown 1`] = `
<styled.div>
  <styled.div>
    <Memo(CartItem)
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
    />
  </styled.div>
  <CartDropDownButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CartDropDownButton>
</styled.div>
`;
